### DISCUSSION TITLE
Fragmented initial state

### DISCUSSION THREAD

**Discussion by tsoos99dev** [pmndrs/zustand#2158] [Q&A] [answered]
Hi. I'm trying to use a Zustand store in a Next.js app. Let's say the app state looks like this:
```typescript
interface Bear {
  id: number;
  name: string;
}

interface Cub {
  id: number;
  name: string;
}

interface AppState {
  bearId?: number;
  bears: Bear[];

  cubId?: number;
  cubs: Cub[]
}
```

The home page has a list of bears, while `/bear/[bearId]` has a list of cubs for the selected bear. The layout file fetches the list of bears on the server in `/` and initialises the store with it according to the example [here](https://docs.pmnd.rs/zustand/guides/initialize-state-with-props). On that route the _bearId_ and _cubId_ remain _undefined_, and the list of cubs stays empty. On `/bear/[bearId]` the list of cubs is collected in the corresponding layout, in addition to the _bearId_ being read from the URL, and I'd like to initialise the store with these.

My problem is that I can't figure out how to initialise the store with these data fragments before the child components start using the store. I tried useEffect, but the children would run their useStore hooks before the data gets into the store. The problem there is that I'd like to avoid handling undefined values, when clearly _bearId_ should be available everywhere in `/bear/[bearId]/**` for example. So this is what the hooks computing the selected bear looks like:

```typescript
export const AppContext = createContext<AppStore | null>(null);

export function useAppStore<T>(selector: (state: AppState) => T): T {
  const store = useContext(AppContext);
  if (!store) throw new Error('Missing AppContext.Provider in the tree');
  return useStore(store, selector);
}

export function useBearOptional() {
  return useAppStore((state) => {
    const bearId = state.bearId;

    if (typeof bearId === 'undefined') return undefined;
    const bear = state.bears.find((b) => b.id === bearId);

    return bear;
  });
}

export function useBear() {
  const bear = useBearOptional();

  if (typeof bear === 'undefined')
    throw new Error("Bear hasn't been provided.");

  return bear;
}
```

If I only needed the selected bear and cub in their respective sub routes, I'd just switch to Context, but I have a navigation bar  displaying the selected bear's and cub's name that lives relatively high in the component tree. This component also displays the list of bears and cubs in a dropdown to ease navigation between arbitrary routes. The names of the bears and cubs are editable, and keeping the nav bar in sync with the page's data is important. The store accomplishes that beautifully, I'm just mentioning that as an argument for the global store.

How could I accomplish the initialisation?

**Comment by dai-shi** [comment]
You can't change the store value during render. useEffect is required.
What can be done is to create a store for each page and initialize it on creation. Multi store solution. But, it's probably something that you don't want.

**Reply by tsoos99dev** [reply]
Yes, thanks for clarifying that useEffect is required. I'd have to somehow sync the stores in that case, so I'd avoid that. Honestly I think I mostly got it working. I used a single store reference passed down to multiple levels of the same provider. I augmented the context with state fragments that the layouts / templates can use to provide additional data at the appropriate places. Then I modified my useStore hook to inject this additional state fragment into the selector call. With the addition of a useEffect call to update the store, and to let go of the state fragments it seems to work quite well. I think it's an iteresting way to combine the localised data fetching in Next.js and the global store on the client side. [Here's](https://github.com/tsoos99dev/next-dynamic-route-state) an example implementation. Maybe someone else also has a similar use case. Hope it could be useful. 

I'd appreciate any feedback you may have on the approach.

**Reply by dai-shi** [reply, marked as answer]
I see. You put both store and the initial value (=fragment) in the context value. I don't think it violates any rules. Good job.

**Reply by tsoos99dev** [reply]
Thank you!
